Programmed soundtrack example

Remi-Godin opened the pull request:
Created soundtrack example, fade-in and fade-out features, added new assets, and updated credits.

# Objective
- Fixes #12651 

## Solution
- Created a resource to hold the track list. 
- The audio assets are then loaded by the asset server and added to the track list. 
- Once the game is in a specific state, an `AudioBundle` is spawned and plays the appropriate track. 
- The audio volume starts at zero and is then incremented gradually until it reaches full volume.
- Once the game state changes, the current track fades out, and a new one fades in at the same time, offering a relatively seamless transition.
- Once a track is completely faded out, it is despawned from the app.
- Game state changes are simulated through a `Timer` for simplicity.
- Track change system is only run if there is a change in the `GameState` resource.
- All tracks are used according to their respective licenses.

alice-i-cecile commented on CREDITS.md:
```diff
@@ -28,6 +28,8 @@
 * FiraMono by The Mozilla Foundation and Telefonica S.A (SIL Open Font License, Version 1.1: assets/fonts/FiraMono-LICENSE)
 * Barycentric from [mk_bary_gltf](https://github.com/komadori/mk_bary_gltf) (MIT OR Apache-2.0)
 * `MorphStressTest.gltf`, [MorphStressTest] ([CC-BY 4.0] by Analytical Graphics, Inc, Model and textures by Ed Mackey)
+* Mysterious accoustic guitar music sample from [florianreichelt](https://freesound.org/people/florianreichelt/sounds/412429/) (CC0 license)
```
We can either fix this typo (from the link), or allow it in our typos config.

alice-i-cecile commented on Cargo.toml:
```diff
@@ -1378,6 +1378,17 @@ description = "Shows how to create and register a custom audio source by impleme
+
+[package.metadata.example.soundtrack]
+name = "Soundtrack"
+description = "Shows how to play different soundtracks"
```
```suggestion
description = "Shows how to play different soundtracks based on game state"
```

alice-i-cecile commented on examples/audio/soundtrack.rs:
```diff
@@ -0,0 +1,154 @@
+//! This example illustrates how to load and play different soundtracks
```
```suggestion
//! This example illustrates how to load and play different soundtracks,
//! transitioning between them as the game state changes.
```

alice-i-cecile commented on examples/audio/soundtrack.rs:
```diff
@@ -0,0 +1,154 @@
+    commands.insert_resource(GameStateTimer(Timer::from_seconds(10.0, TimerMode::Repeating)));
+
+    // Create the track list
+    let track_1 = asset_server.load::<AudioSource>("sounds/Mysterious accoustic guitar.ogg");
```
Yeah, the way this propagates through is unfortunate.

Remi-Godin replied:
What's the best approach for me to fix this? Do I make a suggestion for every line of code that is affected, or do I "Request Changes" ? Still kind of new to contributing to projects, so unsure of how to proceed from here.

alice-i-cecile replied:
As the author, just push changes to your branch that fix the issue.

alice-i-cecile commented on examples/audio/soundtrack.rs:
```diff
@@ -0,0 +1,154 @@
+// Every time the GameState resource changes, this system is run to trigger the song change.
+fn change_track(
+    mut commands: Commands,
+    sountrack_player: Res<SoundtrackPlayer>,
```
Actual typo here :)

Remi-Godin replied:
Oops, my bad!

alice-i-cecile commented on examples/audio/soundtrack.rs:
```diff
@@ -0,0 +1,154 @@
+        .add_plugins(DefaultPlugins)
+        .add_systems(Startup, setup)
+        .add_systems(Update, (cycle_game_state, fade_in, fade_out))
+        .add_systems(Update, change_track.run_if(resource_changed::<GameState>))
```
IMO this run condition is less clear than just checking `.is_changed` inside of `change_track`.

alice-i-cecile replied:
Mildly faster for now, but that's really not worth optimizing. Especially not in a learning example.

bushrat011899 commented on examples/audio/soundtrack.rs:
```diff
@@ -0,0 +1,154 @@
+    }
+}
+
+// Every time the timer ends, switches between the "Peaceful" and "Battle" state.
```
I think it would be nicer to instead react to a user input for the transition, just to really highlight how this is a dynamic system, not just a single audio file with built in fade transitions.

Might also be nice to change the clear colour of the window between a "battle" colour and a "peaceful" colour.

Remi-Godin replied:
I did think about using keyboard input. It is certainly something that can be modified.

I do like the window changing color! But do you think it would stay in the spirit of keeping the examples focused on a single mechanic/system?

alice-i-cecile replied:
Yeah, IMO we should keep it minimal.

bushrat011899 reviewed (approved):
Nice example! Audio is an often overlooked element in game engines, to their detriment.

Remi-Godin commented:
Glad my approach was satisfactory!

Remi-Godin commented:
Apologies, I didn't realize if I pushed to my fork, it was going to reflect on this PR (still new to all this). My intention was not to pollute this PR with unnecessary commits.

I fixed the typo issue and changed the run condition as suggested. However, I forgot to add the comment changes that were suggested to better reflect the purpose of this example.

I also did not change to use keyboard input yet, was waiting for more feedback on it first.

I will be more careful in the future.

alice-i-cecile reviewed (approved):
Once you've regenerated the example text (CI will yell at you), this is good to merge.

SolarLiner commented on examples/audio/soundtrack.rs:
```diff
@@ -0,0 +1,157 @@
+    time: Res<Time>,
+) {
+    for (audio, entity) in audio_sink.iter_mut() {
+        audio.set_volume(audio.volume() + time.delta_seconds() / FADE_TIME);
```
Nothing we can do at this stage for the example but this highlights the need for tweening in the audio engine -- this is going to result in sounding like stepwise motion, because it will only update as fast as the game renders (which is 60 or maybe even 120 Hz which is definitely not fast enough to perceive as gradual).

In this case because the fade is 2 seconds each increment is going to be gradual enough, but shorten the fade time and the problem manifests more clearly.

Remi-Godin replied:
This is also something I considered while programming this. Also, I just did a linear fade on the volume setting, but according to the docs, to decrease the perceived volume by half, you should multiply by 0.316. I figured it was good enough at illustrating the concept while still keeping things simple.

Remi-Godin commented on examples/audio/soundtrack.rs:
```diff
@@ -0,0 +1,157 @@
+            GameState::Peaceful => {
+                commands.spawn((
+                    AudioBundle {
+                        source: soundtrack_player.track_list.first().unwrap().clone(),
+                        settings: PlaybackSettings {
+                            mode: bevy::audio::PlaybackMode::Loop,
+                            volume: bevy::audio::Volume::ZERO,
+                            ..default()
+                        },
+                    },
+                    FadeIn,
+                ));
+            }
+            GameState::Battle => {
+                commands.spawn((
+                    AudioBundle {
+                        source: soundtrack_player.track_list.get(1).unwrap().clone(),
```
I did consider this, but was worried that making it too "static" would not be good for the example, so I decided to go with a `Vec` to show that this is flexible. I almost created a struct to hold not only the track, but a title and volume setting as well since the original issue mentioned configurable volume, but opted against this for simplicity. This is my first example so still not sure as to how in depth they need to be, or how flexible/dynamic they should be. 

I also used `unwrap()` statements instead of doing proper error handling since I wanted to keep the code simple and focused on the subject of the example, and if it cannot get the files, then there's not really any point in running the example and panicking should be fine. I also saw that other examples also uses `unwrap()` statements.

If I do write other examples in the future, I will make sure to keep your feedback in mind. Any other feedback or pointers is also appreciated.
